File: android/StelliveHubAndroid/app/src/main/java/dev/minepacu/stelliveeventnotifier/feature/hubevents/HubEventDetailCalendarPolicy.kt
package dev.minepacu.stelliveeventnotifier.feature.hubevents

import dev.minepacu.stelliveeventnotifier.core.model.HubEvent
import dev.minepacu.stelliveeventnotifier.core.model.HubEventScheduleItem
import dev.minepacu.stelliveeventnotifier.core.model.HubEventScheduleKind
import java.time.LocalDate
import java.time.YearMonth
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

enum class HubEventDetailCalendarMode {
    HIDDEN,
    COMPACT_DATE,
    MONTH_CALENDAR,
}

data class HubEventDetailCalendarDay(
    val date: LocalDate,
    val isInParentEventRange: Boolean,
    val schedules: List<HubEventScheduleItem>,
    val activeScheduleCount: Int,
    val hasDeadline: Boolean,
    val hasOnlyCancelledSchedules: Boolean,
) {
    val scheduleCount: Int
        get() = schedules.size

    val countIndicator: String?
        get() = when (scheduleCount) {
            0 -> null
            1 -> "•"
            in 2..9 -> scheduleCount.toString()
            else -> "9+"
        }
}

data class HubEventDetailCalendarPresentation(
    val mode: HubEventDetailCalendarMode,
    val displayedMonth: YearMonth?,
    val availableMonthRange: ClosedRange<YearMonth>?,
    val selectedDate: LocalDate?,
    val initialSelectedDate: LocalDate?,
    val days: List<HubEventDetailCalendarDay>,
) {
    fun day(date: LocalDate): HubEventDetailCalendarDay? = days.firstOrNull { it.date == date }

    fun scheduleIdsFor(date: LocalDate): List<String> =
        day(date)?.schedules.orEmpty().map(HubEventScheduleItem::id)
}

data class HubEventDetailCalendarHeaderPresentation(
    val title: String,
    val summary: String,
    val scheduleCount: Int,
)

object HubEventDetailCalendarExpansionPolicy {
    fun resolve(
        previousEventId: String?,
        eventId: String,
        currentExpanded: Boolean,
        highlightedScheduleItemId: String?,
    ): Boolean = when {
        previousEventId != eventId -> true
        highlightedScheduleItemId != null -> true
        else -> currentExpanded
    }
}

object HubEventDetailCalendarPolicy {
    val DefaultZoneId: ZoneId = ZoneId.of("Asia/Seoul")

    fun headerPresentation(
        presentation: HubEventDetailCalendarPresentation,
        displayedMonth: YearMonth,
        selectedDate: LocalDate?,
    ): HubEventDetailCalendarHeaderPresentation? {
        if (presentation.mode == HubEventDetailCalendarMode.HIDDEN) return null
        val compactDate = selectedDate ?: presentation.initialSelectedDate
        val relevantDays = when (presentation.mode) {
            HubEventDetailCalendarMode.HIDDEN -> emptyList()
            HubEventDetailCalendarMode.COMPACT_DATE ->
                presentation.days.filter { it.date == compactDate }
            HubEventDetailCalendarMode.MONTH_CALENDAR ->
                presentation.days.filter { YearMonth.from(it.date) == displayedMonth }
        }
        val scheduleCount = relevantDays
            .flatMap(HubEventDetailCalendarDay::schedules)
            .distinctBy(HubEventScheduleItem::id)
            .size
        val hasParentRange = relevantDays.any(HubEventDetailCalendarDay::isInParentEventRange)
        val title = when (presentation.mode) {
            HubEventDetailCalendarMode.HIDDEN -> return null
            HubEventDetailCalendarMode.COMPACT_DATE ->
                detailDateFormatter.format(compactDate ?: return null)
            HubEventDetailCalendarMode.MONTH_CALENDAR -> monthFormatter.format(displayedMonth)
        }
        val summary = when {
            scheduleCount > 0 -> "세부 일정 ${scheduleCount}개"
            presentation.mode == HubEventDetailCalendarMode.COMPACT_DATE -> "행사 일정"
            hasParentRange -> "행사 기간"
            else -> "세부 일정 0개"
        }
        return HubEventDetailCalendarHeaderPresentation(
            title = title,
            summary = summary,
            scheduleCount = scheduleCount,
        )
    }

    fun build(
        event: HubEvent,
        highlightedScheduleItemId: String? = null,
        today: LocalDate = LocalDate.now(DefaultZoneId),
        parentZoneId: ZoneId = DefaultZoneId,
    ): HubEventDetailCalendarPresentation {
        val parentDates = parentDateRange(event, parentZoneId)
        val projectedSchedules = event.scheduleItems
            .sortedWith(
                compareBy<HubEventScheduleItem> { it.startsAt }
                    .thenBy { it.sortOrder }
                    .thenBy { it.id },
            )
            .map { schedule ->
                ProjectedSchedule(
                    schedule = schedule,
                    dates = scheduleDateRange(schedule),
                )
            }
        val allDates = (parentDates + projectedSchedules.flatMap(ProjectedSchedule::dates)).toSortedSet()
        if (allDates.isEmpty()) {
            return HubEventDetailCalendarPresentation(
                mode = HubEventDetailCalendarMode.HIDDEN,
                displayedMonth = null,
                availableMonthRange = null,
                selectedDate = null,
                initialSelectedDate = null,
                days = emptyList(),
            )
        }

        val schedulesByDate = buildMap<LocalDate, MutableList<HubEventScheduleItem>> {
            projectedSchedules.forEach { projected ->
                projected.dates.forEach { date ->
                    getOrPut(date) { mutableListOf() }.apply {
                        if (none { it.id == projected.schedule.id }) add(projected.schedule)
                    }
                }
            }
        }
        val firstDate = allDates.first()
        val lastDate = allDates.last()
        val days = generateSequence(firstDate) { date ->
            date.plusDays(1).takeIf { it <= lastDate }
        }.map { date ->
            val schedules = schedulesByDate[date].orEmpty()
            HubEventDetailCalendarDay(
                date = date,
                isInParentEventRange = date in parentDates,
                schedules = schedules,
                activeScheduleCount = schedules.count { it.cancelledAt == null },
                hasDeadline = schedules.any { it.kind == HubEventScheduleKind.DEADLINE },
                hasOnlyCancelledSchedules = schedules.isNotEmpty() && schedules.all { it.cancelledAt != null },
            )
        }.toList()

        val initialSelectedDate =
            highlightedScheduleItemId
                ?.let { highlightedId -> projectedSchedules.firstOrNull { it.schedule.id == highlightedId } }
                ?.dates
                ?.firstOrNull()
                ?: today.takeIf { date -> date in parentDates || schedulesByDate[date].orEmpty().isNotEmpty() }
                ?: projectedSchedules
                    .asSequence()
                    .filter { it.schedule.cancelledAt == null }
                    .flatMap { projected -> projected.dates.asSequence() }
                    .filter { it >= today }
                    .minOrNull()
                ?: parentDates.firstOrNull()
                ?: projectedSchedules.flatMap(ProjectedSchedule::dates).minOrNull()

        val mode = if (allDates.size == 1) {
            HubEventDetailCalendarMode.COMPACT_DATE
        } else {
            HubEventDetailCalendarMode.MONTH_CALENDAR
        }
        val availableMonthRange = YearMonth.from(firstDate)..YearMonth.from(lastDate)
        return HubEventDetailCalendarPresentation(
            mode = mode,
            displayedMonth = initialSelectedDate?.let(YearMonth::from) ?: YearMonth.from(firstDate),
            availableMonthRange = availableMonthRange,
            selectedDate = initialSelectedDate,
            initialSelectedDate = initialSelectedDate,
            days = days,
        )
    }

    private fun parentDateRange(event: HubEvent, zoneId: ZoneId): List<LocalDate> {
        val start = (event.startsAt ?: event.endsAt)?.atZone(zoneId)?.toLocalDate() ?: return emptyList()
        val requestedEnd = event.endsAt?.atZone(zoneId)?.toLocalDate() ?: start
        return inclusiveDateRange(start, requestedEnd)
    }

    private fun scheduleDateRange(schedule: HubEventScheduleItem): List<LocalDate> {
        val zoneId = runCatching { ZoneId.of(schedule.timezone) }.getOrDefault(DefaultZoneId)
        val start = schedule.startsAt.atZone(zoneId).toLocalDate()
        val requestedEnd = schedule.endsAt?.atZone(zoneId)?.toLocalDate() ?: start
        return inclusiveDateRange(start, requestedEnd)
    }

    private fun inclusiveDateRange(start: LocalDate, requestedEnd: LocalDate): List<LocalDate> {
        val end = requestedEnd.takeIf { it >= start } ?: start
        return generateSequence(start) { date ->
            date.plusDays(1).takeIf { it <= end }
        }.toList()
    }

    private data class ProjectedSchedule(
        val schedule: HubEventScheduleItem,
        val dates: List<LocalDate>,
    )

    private val monthFormatter: DateTimeFormatter =
        DateTimeFormatter.ofPattern("yyyy년 M월", Locale.KOREAN)
    private val detailDateFormatter: DateTimeFormatter =
        DateTimeFormatter.ofPattern("yyyy년 M월 d일 EEEE", Locale.KOREAN)
}

File: ios/StelliveHubiOS/StelliveHubiOS/Views/HubEventDetailCalendarPolicy.swift
import Foundation

enum HubEventDetailCalendarMode: Equatable {
    case hidden
    case compactDate
    case monthCalendar
}

struct HubEventDetailCalendarDate: Hashable, Comparable {
    let year: Int
    let month: Int
    let day: Int

    static func < (lhs: Self, rhs: Self) -> Bool {
        if lhs.year != rhs.year { return lhs.year < rhs.year }
        if lhs.month != rhs.month { return lhs.month < rhs.month }
        return lhs.day < rhs.day
    }

    var calendarMonth: HubEventDetailCalendarMonth {
        HubEventDetailCalendarMonth(year: year, month: month)
    }

    var foundationDate: Date {
        var calendar = HubEventDetailCalendarPolicy.calendar(timeZone: HubEventDetailCalendarPolicy.fallbackTimeZone)
        return calendar.date(from: DateComponents(year: year, month: month, day: day))!
    }

    func addingDays(_ value: Int) -> Self {
        var calendar = HubEventDetailCalendarPolicy.calendar(timeZone: HubEventDetailCalendarPolicy.fallbackTimeZone)
        let result = calendar.date(byAdding: .day, value: value, to: foundationDate)!
        return HubEventDetailCalendarPolicy.localDate(result, timeZone: calendar.timeZone)
    }
}

struct HubEventDetailCalendarMonth: Hashable, Comparable {
    let year: Int
    let month: Int

    static func < (lhs: Self, rhs: Self) -> Bool {
        lhs.year == rhs.year ? lhs.month < rhs.month : lhs.year < rhs.year
    }

    func addingMonths(_ value: Int) -> Self {
        let monthIndex = year * 12 + (month - 1) + value
        return Self(year: monthIndex / 12, month: monthIndex % 12 + 1)
    }
}

struct HubEventDetailCalendarDay: Equatable {
    let date: HubEventDetailCalendarDate
    let isInParentEventRange: Bool
    let schedules: [HubEventScheduleItem]

    var activeScheduleCount: Int {
        schedules.lazy.filter { $0.cancelledAt == nil }.count
    }

    var scheduleCount: Int {
        schedules.count
    }

    var hasDeadline: Bool {
        schedules.contains { $0.kind == .deadline && $0.cancelledAt == nil }
    }

    var hasOnlyCancelledSchedules: Bool {
        !schedules.isEmpty && activeScheduleCount == 0
    }

    var badgeText: String? {
        switch scheduleCount {
        case 0, 1: nil
        case 2...9: String(scheduleCount)
        default: "9+"
        }
    }

    var hasContent: Bool {
        isInParentEventRange || !schedules.isEmpty
    }

    var accessibilityDescription: String {
        var components = ["\(date.year)년 \(date.month)월 \(date.day)일"]
        if isInParentEventRange { components.append("행사 기간") }
        if scheduleCount > 0 { components.append("일정 \(scheduleCount)개") }
        if hasOnlyCancelledSchedules { components.append("모두 취소됨") }
        if hasDeadline { components.append("마감 일정 포함") }
        return components.joined(separator: ", ")
    }
}

struct HubEventDetailCalendarPresentation: Equatable {
    let mode: HubEventDetailCalendarMode
    let displayedMonth: HubEventDetailCalendarMonth?
    let availableMonthRange: ClosedRange<HubEventDetailCalendarMonth>?
    let selectedDate: HubEventDetailCalendarDate?
    let initialSelectedDate: HubEventDetailCalendarDate?
    let today: HubEventDetailCalendarDate
    let days: [HubEventDetailCalendarDate: HubEventDetailCalendarDay]
    let earliestDate: HubEventDetailCalendarDate?
    let latestDate: HubEventDetailCalendarDate?

    func day(for date: HubEventDetailCalendarDate) -> HubEventDetailCalendarDay {
        days[date] ?? HubEventDetailCalendarDay(
            date: date,
            isInParentEventRange: false,
            schedules: []
        )
    }

    func scheduleIDs(on date: HubEventDetailCalendarDate) -> [String] {
        day(for: date).schedules.map(\.id)
    }

    func monthGrid(for month: HubEventDetailCalendarMonth) -> [HubEventDetailCalendarDate] {
        let first = HubEventDetailCalendarDate(year: month.year, month: month.month, day: 1)
        var calendar = HubEventDetailCalendarPolicy.calendar(timeZone: HubEventDetailCalendarPolicy.fallbackTimeZone)
        let weekday = calendar.component(.weekday, from: first.foundationDate)
        let gridStart = first.addingDays(-(weekday - 1))
        return (0..<42).map { gridStart.addingDays($0) }
    }

    func contains(_ date: HubEventDetailCalendarDate) -> Bool {
        guard let earliestDate, let latestDate else { return false }
        return date >= earliestDate && date <= latestDate
    }
}

enum HubEventDetailCalendarPolicy {
    static let fallbackTimeZone = TimeZone(identifier: "Asia/Seoul")!

    private struct ScheduleProjection {
        let item: HubEventScheduleItem
        let dates: [HubEventDetailCalendarDate]
    }

    static func presentation(
        for event: HubEvent,
        highlightedScheduleItemID: String? = nil,
        now: Date = Date()
    ) -> HubEventDetailCalendarPresentation {
        let parentDates = projectedDates(
            startsAt: event.startsAt,
            endsAt: event.endsAt,
            timeZone: fallbackTimeZone
        )
        let parentDateSet = Set(parentDates)
        let scheduleProjections = projectedSchedules(event.scheduleItems)
        var schedulesByDate: [HubEventDetailCalendarDate: [HubEventScheduleItem]] = [:]

        for projection in scheduleProjections {
            for date in projection.dates {
                if schedulesByDate[date, default: []].contains(where: { $0.id == projection.item.id }) {
                    continue
                }
                schedulesByDate[date, default: []].append(projection.item)
            }
        }

        let relevantDates = parentDateSet.union(schedulesByDate.keys).sorted()
        let today = localDate(now, timeZone: fallbackTimeZone)
        guard let earliestDate = relevantDates.first, let latestDate = relevantDates.last else {
            return HubEventDetailCalendarPresentation(
                mode: .hidden,
                displayedMonth: nil,
                availableMonthRange: nil,
                selectedDate: nil,
                initialSelectedDate: nil,
                today: today,
                days: [:],
                earliestDate: nil,
                latestDate: nil
            )
        }

        var days: [HubEventDetailCalendarDate: HubEventDetailCalendarDay] = [:]
        for date in relevantDates {
            days[date] = HubEventDetailCalendarDay(
                date: date,
                isInParentEventRange: parentDateSet.contains(date),
                schedules: schedulesByDate[date] ?? []
            )
        }

        let initialSelectedDate = initialSelection(
            highlightedScheduleItemID: highlightedScheduleItemID,
            today: today,
            parentDates: parentDates,
            schedules: scheduleProjections
        ) ?? earliestDate
        let firstMonth = earliestDate.calendarMonth
        let lastMonth = latestDate.calendarMonth
        let mode: HubEventDetailCalendarMode = relevantDates.count == 1
            ? .compactDate
            : .monthCalendar

        return HubEventDetailCalendarPresentation(
            mode: mode,
            displayedMonth: initialSelectedDate.calendarMonth,
            availableMonthRange: firstMonth...lastMonth,
            selectedDate: initialSelectedDate,
            initialSelectedDate: initialSelectedDate,
            today: today,
            days: days,
            earliestDate: earliestDate,
            latestDate: latestDate
        )
    }

    static func calendar(timeZone: TimeZone) -> Calendar {
        var calendar = Calendar(identifier: .gregorian)
        calendar.locale = Locale(identifier: "ko_KR")
        calendar.timeZone = timeZone
        calendar.firstWeekday = 1
        return calendar
    }

    static func localDate(_ date: Date, timeZone: TimeZone) -> HubEventDetailCalendarDate {
        let components = calendar(timeZone: timeZone).dateComponents([.year, .month, .day], from: date)
        return HubEventDetailCalendarDate(
            year: components.year!,
            month: components.month!,
            day: components.day!
        )
    }

    static func resolvedTimeZone(for item: HubEventScheduleItem) -> TimeZone {
        TimeZone(identifier: item.timezone) ?? fallbackTimeZone
    }

    private static func projectedSchedules(
        _ items: [HubEventScheduleItem]
    ) -> [ScheduleProjection] {
        var seenIDs = Set<String>()
        return items
            .sorted(by: scheduleSort)
            .compactMap { item in
                guard seenIDs.insert(item.id).inserted else { return nil }
                return ScheduleProjection(
                    item: item,
                    dates: projectedDates(
                        startsAt: item.startsAt,
                        endsAt: item.endsAt,
                        timeZone: resolvedTimeZone(for: item)
                    )
                )
            }
    }

    private static func projectedDates(
        startsAt: Date?,
        endsAt: Date?,
        timeZone: TimeZone
    ) -> [HubEventDetailCalendarDate] {
        guard let startsAt else {
            return endsAt.map { [localDate($0, timeZone: timeZone)] } ?? []
        }
        let start = localDate(startsAt, timeZone: timeZone)
        guard let endsAt else { return [start] }
        let end = localDate(endsAt, timeZone: timeZone)
        guard end >= start else { return [start] }

        var result: [HubEventDetailCalendarDate] = []
        var date = start
        while date <= end {
            result.append(date)
            date = date.addingDays(1)
        }
        return result
    }

    private static func initialSelection(
        highlightedScheduleItemID: String?,
        today: HubEventDetailCalendarDate,
        parentDates: [HubEventDetailCalendarDate],
        schedules: [ScheduleProjection]
    ) -> HubEventDetailCalendarDate? {
        if let highlightedScheduleItemID,
           let highlighted = schedules.first(where: { $0.item.id == highlightedScheduleItemID }),
           let date = highlighted.dates.first {
            return date
        }

        if parentDates.contains(today) || schedules.contains(where: { $0.dates.contains(today) }) {
            return today
        }

        if let nextDate = schedules
            .filter({ $0.item.cancelledAt == nil })
            .flatMap(\.dates)
            .filter({ $0 > today })
            .min() {
            return nextDate
        }

        if let parentStart = parentDates.first {
            return parentStart
        }

        return schedules.flatMap(\.dates).min()
    }

    private static func scheduleSort(
        _ left: HubEventScheduleItem,
        _ right: HubEventScheduleItem
    ) -> Bool {
        if left.startsAt != right.startsAt { return left.startsAt < right.startsAt }
        if left.sortOrder != right.sortOrder { return left.sortOrder < right.sortOrder }
        return left.id < right.id
    }
}
